### PR TITLE
Initial pre-commit config

### PR DESCRIPTION
**Pull request by nicolas-guichard**
Following the discussion in https://github.com/mozsearch/mozsearch/pull/839, I've put together a .pre-commit-config.yaml with clang-format, rustfmt, clippy and a few extra hooks and made sure that it passes.

To use it:
- install pre-commit (using your preferred package manager)
- run `pre-commit install` to install the hook

When committing those checks will be run on the files that are touched by the commit. You can also run the checks manually using `pre-commit run --all-files`.

It does add a bit of overhead when committing, especially the cargo-clippy hook from a cold cache:
```
➜  mozsearch git:(wip-stack) rm -rf **/target

➜  mozsearch git:(wip-stack) time pre-commit run --all-files                                                                                                                                   
trim trailing whitespace.................................................Passed
fix end of files.........................................................Passed
check for added large files..............................................Passed
check for case conflicts.................................................Passed
check xml................................................................Passed
check yaml...............................................................Passed
check json...............................................................Passed
check for broken symlinks............................(no files to check)Skipped
detect destroyed symlinks................................................Passed
check that executables have shebangs.....................................Passed
clang-format.............................................................Passed
cargo fmt................................................................Passed
cargo clippy.............................................................Passed
pre-commit run --all-files  407,98s user 59,95s system 1163% cpu 40,222 total

➜  mozsearch git:(wip-stack) time pre-commit run --all-files
trim trailing whitespace.................................................Passed
fix end of files.........................................................Passed
check for added large files..............................................Passed
check for case conflicts.................................................Passed
check xml................................................................Passed
check yaml...............................................................Passed
check json...............................................................Passed
check for broken symlinks............................(no files to check)Skipped
detect destroyed symlinks................................................Passed
check that executables have shebangs.....................................Passed
clang-format.............................................................Passed
cargo fmt................................................................Passed
cargo clippy.............................................................Passed
pre-commit run --all-files  7,97s user 1,81s system 327% cpu 2,982 total
```

### CODE REVIEW

File: sax/sax.patch
Comment (asutherland) [inline sax/sax.patch:1]: I think we need to exclude this file from being reformatted.

**Review by asutherland — APPROVED** [review]
This is great!  When running locally I saw some clippy errors so I did a `rustup update` and now I see a different error:

```
   Compiling tools v0.1.0 (/home/visbrero/rev_control/fgit/mozsearch-alt/tools)
error: this `map_or` is redundant
   --> src/format.rs:755:36
    |
755 |             let same_rev_as_last = last_revs.map_or(false, |last| last == revs);
    |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use a standard comparison instead: `(last_revs == Some(revs))`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_map_or
    = note: `-D clippy::unnecessary-map-or` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::unnecessary_map_or)]`

error: could not compile `tools` (lib) due to 1 previous error
`cargo clippy` found errors
```

rs=asuth for fixing this any/similar errors without additional review.

File: sax/sax.patch
Comment (nicolas-guichard) [reply sax/sax.patch:1]: Indeed, I've disabled the trailing-whitespace hook for all `.patch` files.

**Comment by nicolas-guichard** [comment]
What do you mean with `rs=`?
I understand `r=` and `r?` from https://firefox-source-docs.mozilla.org/contributing/reviews.html#commit-message-syntax but `rs` is new to me.

**Comment by asutherland** [comment]
> What do you mean with `rs=`? I understand `r=` and `r?` from https://firefox-source-docs.mozilla.org/contributing/reviews.html#commit-message-syntax but `rs` is new to me.

Ah, sorry, it's a very old Mozilla-ism for "rubber stamp" that is referenced [here in the context of the now obsolete super-review concept mentioned in the points above it](https://www-archive.mozilla.org/hacking/code-review-faq#10).  With super-review no longer being a thing, my use, and I think sort of the convention when people use it is: I agree with the concept of this proposed fix and I think the fix itself is likely to be straightforward/mechanical.

In other words, it's maybe along the lines of the :100: emoji these days but saying, seriously, I agree with this plan and r+.

In this case I should have just said `r=asuth` as I think the rest of the prose was otherwise clear without the `rs=` thing no one should need to know.
